File: 5Days-SoftwareArchitecture/typesript/Day1- Metrics/WMCExample.ts
// POOR: High WMC
class Processor {
    process(user: string, amount: number): void {
        if (user && user.length > 0) {
            if (amount > 0) {
                console.log("Valid user and amount, processing...");
                if (amount < 1000) console.log("Small payment, fast track");
                else console.log("Large payment, require approval");
            } else {
                console.log("Invalid amount!");
            }
        } else {
            console.log("Invalid user!");
        }
    }
}

// GOOD: Low WMC
class CleanProcessor {
    process(user: string, amount: number): void {
        if (!this.validateUser(user)) return;
        if (!this.validateAmount(amount)) return;
        this.handlePayment(amount);
    }

    private validateUser(user: string): boolean {
        if (!user || user.length === 0) {
            console.log("Invalid user!");
            return false;
        }
        return true;
    }

    private validateAmount(amount: number): boolean {
        if (amount <= 0) {
            console.log("Invalid amount!");
            return false;
        }
        return true;
    }

    private handlePayment(amount: number): void {
        console.log("Processing payment:", amount);
        if (amount < 1000) console.log("Small payment, fast track");
        else console.log("Large payment, require approval");
    }
}

// Usage
new Processor().process("Alice", 500);
new CleanProcessor().process("Bob", 1500);

File: 5Days-SoftwareArchitecture/Java/Day1 CC Metrics/WMCExample.java
// POOR: High WMC
class Processor {
    void process(String user, double amount) {
        if (user != null && !user.isEmpty()) {
            if (amount > 0) {
                System.out.println("Valid user and amount, processing...");
                if (amount < 1000) {
                    System.out.println("Small payment, fast track");
                } else {
                    System.out.println("Large payment, require approval");
                }
            } else {
                System.out.println("Invalid amount!");
            }
        } else {
            System.out.println("Invalid user!");
        }
    }
}

// GOOD: Low WMC (Split logic into multiple methods)
class CleanProcessor {
    void process(String user, double amount) {
        if (!validateUser(user)) return;
        if (!validateAmount(amount)) return;
        handlePayment(amount);
    }

    private boolean validateUser(String user) {
        if (user == null || user.isEmpty()) {
            System.out.println("Invalid user!");
            return false;
        }
        return true;
    }

    private boolean validateAmount(double amount) {
        if (amount <= 0) {
            System.out.println("Invalid amount!");
            return false;
        }
        return true;
    }

    private void handlePayment(double amount) {
        System.out.println("Processing payment: " + amount);
        if (amount < 1000) System.out.println("Small payment, fast track");
        else System.out.println("Large payment, require approval");
    }
}

// Usage
public class WMCExample {
    public static void main(String[] args) {
        new Processor().process("Alice", 500);
        new CleanProcessor().process("Bob", 1500);
    }
}
